Add protos that define API between client and storage node

amytai opened the pull request.

amytai committed:
Add protos that define API between client and storage node

no2chem commented:
Comments please so I can figure out what's going on :)

no2chem commented on storage.proto:
```diff
@@ -0,0 +1,67 @@
+}
+
+//TODO: perhaps coalesce Balance/AccountExists/GetStorage requests into a single request in the future
+message BalanceRequest {
```
Should be called `ValueRequest`

amytai replied:
Updated to `AccountRequest`

no2chem commented on storage.proto:
```diff
@@ -0,0 +1,67 @@
+  bytes address = 1;
+}
+
+message BalanceReply {
```
Same as above

amytai replied:
Rolled into `AccountRequest`

no2chem commented on storage.proto:
```diff
@@ -0,0 +1,67 @@
+  bool hasCode = 1;
+  bytes code = 2;
+  uint64 size = 3; // Should be of type size_t natively
+  bytes hash = 4; // Should be 32 bytes long, of type keccak256 hash. If no code, then hash of empty code.
```
If we're interested in saving bandwidth, the client should compute the hash

amytai replied:
Yeah this has been modified; code hash is rolled into `AccountRequest` proto, which the client will decode.

no2chem commented on storage.proto:
```diff
@@ -0,0 +1,67 @@
+  rpc getBlockHash (BlockHashRequest) returns (BlockHashReply) {}
+}
+
+message CodeRequest {
```
Maybe we could simplify this, just retrieve the code+size

amytai replied:
Don't even think you need to retrieve the size, since the client can determine that locally, since I am forcing the storage node to send the entire code block.

no2chem commented on storage.proto:
```diff
@@ -0,0 +1,67 @@
+
+message AccountExistsReply {
+  bool exists = 1;
+  bytes proof = 2;
```
Hm, this will require a non-existence proof. Do we have support for that, @SoujanyaPonnapalli ?

amytai replied:
ping on this. For now, I'm assuming there's a non-existence proof (we will need one).

SoujanyaPonnapalli replied:
We don't support non-existence proofs as yet. I will work on that!

no2chem commented on storage.proto:
```diff
@@ -0,0 +1,67 @@
+}
+
+message BlockHashRequest {
+  int64 number = 1;
```
Per the other discussion, I think we want a `mostRecentHashes` request of some sort.

amytai replied:
`mostRecentHashes` can be implemented in this (more general) RPC call, right? Basically you would send a `BlockHashRequest{.number = -1}` (where number == -1 indicates you want most recent), and then the `BlockHashReply` gives you a list of block hashes.
In particular, `BlockHashReply` is a list of block hashes, which allows the Storage Node to send multiple blocks, if there are unresolved forks for the block number requested.

no2chem reviewed (commented):
see comments

amytai commented:
Added comments and modified the interface quite a bit. But I think it's now in much better shape and can actually handle the witnesses returned by the MPT.
Still need to figure out how to do RPCs for the get/set_storage calls (probably don't even need one for set_storage..), but that should be it.

SoujanyaPonnapalli reviewed (approved):
Looks good to me.

proof in RPCWitness could just be repeated bytes?

amytai commented:
@SoujanyaPonnapalli Yep, I thought about that too. But I'm going to leave it for now, since the additional abstraction makes it obvious what the field is for. Will consider inlining it as you described in the future
